LIMS2 NEW HUMAN ASSEMBLY WORK
-----------------------------

GRCh37 ( hg19 ) -> GRCh38 ( hg38 )

Work must be coordinated with WGE liftover as the two systems share code / interfaces:
- Design Creation
- Design Import
- Crispr Import

# TASKS 
Items marked ** can be tested in dev but need to be carried out in the live environment just before deployment

- Liftover coordinates from old to new assembly ( see liftover section ) **
    - design oligos
    - genotyping primers
    - crisprs
    - crispr primers
- Set new species default assembly in `species_default_assembly` table **
- Insert new assembly row in Assemblies table **
- Install new version of Ensembl API and point config files to it
    - New ensembl api: /opt/t87/global/software/ensembl/ensembl-core-[version]
    - config: /opt/t87/global/conf/bashrc.d/ensembl **
- UCSC Blat links need to assembly name
    - LIMS2::Constants : `%UCSC_BLAT_DB` UCSC version names ( e.g. hg19 to hg38 )
- Design Creation *LIMS2 && WGE*
    - new bwa aln ref file
        - point to srpipe files
    - new aos ref chromosome files
       - downloaded from Ensembl, one file for each chromosome
       - cleaned up for aos ( fasta header just chromosome name, nothing else )
    - point to new files above:
        - DesignCreate::Constants: %BWA_GENOME_FILES && %DEFAULT_CHROMOSOME_DIR
    - new default assembly
        - DesignCreate::Constants: %CURRENT_ASSEMBLY,
    - farm3 design create env setup script needs changing **
        - `run_in_farm3`
        - point to new ensembl api
- Crispr ES QC
    - new bwa mem & samtools index files
        - look in /lustre/scratch109/blastdb/Users/team87/Human/bwa/README.txt
    - vep changes ( ensembl tools v76 )
        - upgraded because the ensembl cache we use only has GRCh38 for v76
        - see http://www.ensembl.org/info/docs/tools/vep/script/vep_download.html for install details
        - when using install script do not download cache or fasta files
    - modified qc viewer to use crispr coordiantes stored on well qc result ( not direct from crispr object, wrong assembly for old runs )
- EngSeqBuilder - changed default human assembly in EnsSeqBuilder.pm %ASSEMBLY hash
- Design Targets, regenerate for new assembly & build
    - download latest protein coding human gene list:
        - http://www.genenames.org/cgi-bin/statistics - protein-coding genes -> custom download
        - Pick follwing fields: HGNC ID, Approved Symbol, Ensembl Gene ID & Ensembl ID ( supplied by Ensembl )
        - Submit, save generated tsv file
        - clean up file for next step ( csv not tsv, column headers changed )
    - run `DesignCreate/script/gibson_design_targets.pl`
        - first run: input file is list of protein coding genes, use --strict criteria
        - second run: input is failed genes from last run, do not use --strict this time
    - transfer over any manually picked targets from old assembly
        - wrote a script to help with this
        - DesignCreate/script/transfer_manually_picked_targets_between_assemblies.pl
        - must specify species and old assembly id
        - output is target_parameters.csv file
    - Load new design targets into LIMS2 **
        - merge the multiple target_parameters.csv files from the last 2 steps into one
        - convert the csv file into yaml
        - Use lims2-task load-design-targets with the merged target_parameters.yaml file
- Genoverse - David did the following work
    - A new metadata file for the grch38 assembly was required for Genoverse.
    - This is installed in WebApp-Common: `shared_static/Genoverse/js/genomes/grch38.js`
    - The file is generated by running a slightly modified node.js file (same location).
    - I created a new js file called `make_cgrch38.js` in ~dp10/work/genoverse. I have an installation of node and required support modules in my user area (look in ~dp10/bin/bin/node ). The command I ran to generate the new metadata file was:
      `~/bin/bin/node ./make_GRCh38.js > grch38.js`
    - I then edited the top of the file to assign the correct variable for genoverse to use ( var grch38 = ...).
    - LIMS2: `root/site/users/crisprs` - the three genoverse TT files needed occurrences of grch37 changing to grch38 and move from beta.rest to rest.ensembl.
    - The OligoSelection.pm module in LIMS2 need minor updates to correctly return data for a single assembly. 
- QC Template Eng Seq Params
    - update eng seq params in the qc template wells to the new coordinates for a given design **
    - wrote script to do this: `LIMS2::Scripts::liftover::qc_templates.pl`
    - do not need to update crispr vector template plates ( the crispr sequence is stored as part of the eng seq params )

## One Off Work
- WGE to LIMS2 Design and Crispr Import
    - Stopped import if the design to import is not the default species assembly on LIMS2
- Crispr Import - no assembly
    - same logic added for crisprs and crispr pairs
    - wait till changes made in WGE so we can get Assembly of a crispr
- Store assembly in crispr es qc well analysis json 

# TEST FOLLOWING
- look at new coordinates for:
    - design oligos
    - crisprs
    - genotyping primers
    - crispr primers
- UCSC Blat of design oligos
- Design Target - report
- Design Creation - new human design
- Eng Seq Builder
    - check genbank file creation same on 2 different assemblies
    - Can use EngSeqBuilder/bin/lims2-genbank.pl script
- Crispr ES QC
- Genotyping Primers
    - Davids primer generation code 
    - crispr group primers
- Genoverse Browser
- WGE Crispr Importer
- WGE Design Importer

-----------------------------

# UCSC Liftover Tool
Work out new assembly coordinates given the old assembly coordinates.
Download from: `http://hgdownload.cse.ucsc.edu/admin/exe/linux.x86_64.v287/`
Grab chain file for our conversion from:`http://hgdownload.cse.ucsc.edu/goldenPath/hg19/liftOver/`
- hg19ToHg38.over.chain.gz

## Liftover Tool Steps
- Grab current coordinates for species we are lifting over
- Create a bed file of this data, the id of each oligo should be something we can use to insert data back into LIMS2
    - e.g. design oligo id
- Grab correct chain file from UCSC, need to be for the right species and the right transition
    - e.g. mm9ToMm10.over.chain
- run liftover tool on bed file with the chain file
    - `liftOver input.bed chain_file output.bed unmatched.bed`
- check the unmatched.bed file for any liftover failures ( fix if you can )
- Insert the new oligo coordinates from the output.bed file into LIMS2

-----------------------------

# Current Deployment Notes

# RELEASE
- LIMS2::WebApp ( `new_human_assembly` branch )
- Design::Create ( `new_human_assembly` branch ) - merged to devel
- HTGT::QC::Common ( `new_human_assembly` branch )
- EngSeqBuilder  ( `new_human_assembly` branch )
- WebAppCommon ( `new_human_assembly` branch )
- LIMS2::Util  ( `new_human_assembly` branch )
- WGE ( `grch38` branch )

# DELETE AFTER
- ensembl-tools-75
- old ensembl api versions
